python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -1175,20 +1171,77 @@ class PyInferenceRequest
     ThrowIfError(TRITONSERVER_InferenceRequestSetDoubleParameter(
         triton_object_, key.c_str(), value));
   }
+
   void Cancel()
   {
     ThrowIfError(TRITONSERVER_InferenceRequestCancel(triton_object_));
   }
 
+  // Response management
+  void GetNextResponse(const py::object& py_future)
+  {
+    std::lock_guard lock(response_mu_);
 
- public:
-  std::unique_ptr<CallbackResource> request_callback_resource_{nullptr};
+    if (responses_.empty()) {
+      if (response_future_.get() != nullptr) {
+        throw AlreadyExistsError("cannot call GetNextResponse concurrently");
+      }
+      response_future_.reset(new py::object(py_future));
+    } else {
+      std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t>&
+          py_response = responses_.front();
+      PyFutureSetResult(py_future, py_response);
+      responses_.pop_front();
+    }
+  }
 
  private:
-  std::unique_ptr<PyResponseAllocator::CallbackResource>
-      allocator_callback_resource_{nullptr};
-  std::unique_ptr<PyInferenceResponse::CallbackResource>
-      response_callback_resource_{nullptr};
+  void AddNextResponse(
+      struct TRITONSERVER_InferenceResponse* response, const uint32_t flags)
+  {
+    std::shared_ptr<PyInferenceResponse> managed_ptr(nullptr);
+    if (response != nullptr) {
+      managed_ptr.reset(new PyInferenceResponse(response, true /* owned */));
+    }
+    std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t> py_response(
+        std::move(managed_ptr), std::move(flags));
+    {
+      std::lock_guard lock(response_mu_);
+      if (response_future_.get() == nullptr) {
+        responses_.push_back(std::move(py_response));
+        return;
+      }
+    }
+    {
+      py::gil_scoped_acquire gil;
+      std::unique_ptr<py::object> response_future_local(nullptr);
+      response_future_.swap(response_future_local);
```

Comment:
not sure I follow this  - we create a new py object from null, swap it with the stored one (releasing it?) - then set result on the nullptr based object?...

Or ... I guess swap works the opposite way :) 

we swap the null ptr for the one stored in the future ....

Reply:
the guarantee here that the `GetNextResponse()` will not be called and receive a new future object (vs this future object) is this future object is not yet set with a result, so as soon as this future object is set with a result, we need to be ready for the next future object from `GetNextResponse()` that may take the place of this future object. thus, the solution here is to move the current future object from the class to local variable, making the class variable ready for the next future object before the current future object is set with a result.